fedbiomed/researcher/federated_workflows/_experiment.py

```diff
@@ -847,36 +865,159 @@ def run_once(self, increase: bool = False, test_after: bool = False) -> int:
-        if self._agg_optimizer is not None:
-            self._agg_optimizer.set_aux(aux_var)
-        # If no Optimizer is used but auxiliary variables were received, raise.
-        elif aux_var:
+        failed = []  # type: List[str]
+        auxvar = {}  # type: Dict[str, AuxVar]
+        for node_name, node_dict in nodes_auxvar.items():
+            if not isinstance(node_dict, dict):
+                failed.append(node_name)
+            elif not failed:
+                for mod_name, mod_auxv in node_dict.items():
+                    auxvar[mod_name] = auxvar.get(mod_name, 0) + mod_auxv
+        if failed:
             raise FedbiomedExperimentError(
-                "Received auxiliary variables from 1+ node Optimizer, but "
-                "no `agg_optimizer` was set for this Experiment to process "
-                "them.\nThese variables come from the following plug-in "
-                f"modules: {set(aux_var)}."
+                "Received unproper-type optimizer auxiliary variables from "
+                "1+ nodes.\nNodes that sent invalid data are the following: "
+                f"{failed}."
+            )
+        return auxvar
+
+    def _aggregate_encrypted_model_params_and_optim_auxvar(
+        self,
+        model_params: Dict[str, List[int]],
+        optim_auxvar: Dict[str, EncryptedAuxVar],
+        encryption_factors: Optional[Dict[str, List[int]]],
+        total_sample_size: int,
+    ) -> Tuple[
+        Dict[str, Union[np.ndarray, torch.Tensor]],
+        Optional[Dict[str, AuxVar]],
+    ]:
+        """Secure-aggregate nodes' model parameters and (opt.) optimizer auxiliary variables.
+
+        Args:
+            model_params: Dict of node-wise flattened encrypted model parameters.
+            optim_auxvar: Dict of node-wise encrypted optimizer auxiliary variables.
+                May be empty.
+            encryption_factors: Optional dict of node-wise encryption factors, used
+                to verify that SecAgg was properly conducted.
+            total_sample_size: Total number of training samples used by nodes.
+
+        Returns:
+            aggregated_params: Aggregated model parameters, unflattened into a dict
+                mapping parameters' names to their values.
+            aggregated_auxvar: Optional aggregated optimizer auxiliary variables,
+                unflattened into a dict mapping modules' names to `AuxVar` objects
+                they are meant to receive and process.
+        """
+        # Gather parameters that need aggregate-decryption.
+        encrypted = model_params
+        encrypted_auxvar = None  # type: Optional[EncryptedAuxVar]
+        aggregated_auxvar = None  # type: Optional[Dict[str, AuxVar]]
+        n_aux_var = 0
+
+        if optim_auxvar:
+            # Ensure auxiliary variables have the same node-order as model
+            # parameters, type-check and pre-aggregate them.
+            encrypted_auxvar = self._preaggregate_encrypted_optim_auxvar(
```

Comment:
I fail to understand the whole process here. So first auxvar are preaggregated and then  aggregated ? 
What is the difference here?
It is really confusing and not very straightforward in my opinion